bundle/edu.gemini.shared.util/src/main/java/edu/gemini/shared/util/immutable/Either.java

```diff
@@ -0,0 +1,80 @@
+package edu.gemini.shared.util.immutable;
+
+import java.io.Serializable;
+
+/**
+ * Simulates a Scala Either[S,T].
+ */
+public final class Either<L,R> implements Serializable {
+    private final Option<L> left;
+    private final Option<R> right;
+
+    public static <L,R> Either<L,R> left(final L left) {
+        return new Either<>(new Some<>(left), None.instance());
+    }
+
+    public static <L,R> Either<L,R> right(final R right) {
+        return new Either<>(None.instance(), new Some<>(right));
+    }
+
+    private Either(final Option<L> left, final Option<R> right) {
+        this.left  = left;
+        this.right = right;
+    }
+
+    public Option<L> getLeft() {
+        return left;
+    }
+
+    public Option<R> getRight() {
+        return right;
+    }
+
+    public <T> T fold(final Function1<? super L, ? extends T> leftFunc,
+                      final Function1<? super R, ? extends T> rightFunc) {
+
+        return left.map(leftFunc).getOrElse(() -> right.map(rightFunc).getValue());
+    }
+
+    public <T> Option<T> foldLeft(final Function1<? super L, ? extends T> leftFunc) {
+        return left.map(leftFunc);
+    }
+
+    public <T> Option<T> foldRight(final Function1<? super R, ? extends T> rightFunc) {
+        return right.map(rightFunc);
+    }
```

Comment:
I would expect this to work like `scalaz.\/` and take a zero value of type `Z` and a function that combines a `Z` and an `R`.  The zero value is just returned if left but otherwise the function is applied on that and the right value.